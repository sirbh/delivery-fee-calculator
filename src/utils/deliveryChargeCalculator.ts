import { FeeCalculatorConstants } from './appConstants';

const {
    bulkItemFee,
    extraItemFee,
    freeDeliveryCartValue,
    maxRushHour,
    minCartValue,
    minDistance,
    minDistanceFee,
    minItemAmount,
    minRushHour,
    rushHourDay,
    stepDistance,
    stepDistanceFee,
    bulkItemNumber,
    rushHourMultiplier,
} = FeeCalculatorConstants;

/**
 * Calculates delivery fee
 * @param {number} cartValue cart price in format 00.00 euro
 * @param {number} deliveryDistance distance as an integer
 * @param {number} numberOfItems total items in cart
 * @param {number} time time of delivery
 * @returns {number} delivery fee in cents
 */

function deliveryChargeCalculator(
    cartValue: number,
    deliveryDistance: number,
    numberOfItems: number,
    time: Date
): number {
    const cartValueInCents = Math.round(cartValue * 100);
    let deliveryFeeInCents = 0;

    if (cartValueInCents >= freeDeliveryCartValue) {
        return deliveryFeeInCents;
    }

    if (cartValueInCents < minCartValue) {
        deliveryFeeInCents += minCartValue - cartValueInCents;
    }

    if (deliveryDistance > 0) {
        deliveryFeeInCents += minDistanceFee;
    }

    if (deliveryDistance > minDistance) {
        const extraDistance = deliveryDistance - minDistance;
        deliveryFeeInCents +=
            Math.ceil(extraDistance / stepDistance) * stepDistanceFee;
    }

    if (numberOfItems > minItemAmount) {
        const extraItems = numberOfItems - minItemAmount;
        deliveryFeeInCents += extraItems * extraItemFee;

        if (numberOfItems > bulkItemNumber) {
            deliveryFeeInCents += bulkItemFee;
        }
    }

    if (
        time.getDay() === rushHourDay &&
        time.getHours() <= maxRushHour &&
        time.getHours() >= minRushHour
    ) {
        deliveryFeeInCents = Math.round(
            deliveryFeeInCents * rushHourMultiplier
        );
    }

    if (deliveryFeeInCents > 1500) {
        return NaN;
    }

    return deliveryFeeInCents;
}

export default deliveryChargeCalculator;
