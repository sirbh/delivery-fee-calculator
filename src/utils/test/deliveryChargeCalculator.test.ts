import { describe, expect, it } from 'vitest';
import deliveryChargeCalculator from '../deliveryChargeCalculator';

describe('delivery charge calculator', () => {
    it('should calulate surcharge for cart value less than 10 euro', () => {
        expect(
            deliveryChargeCalculator(8.9, 100, 1, new Date(2022, 1, 2, 14, 30))
        ).toBe(310);
    });

    it('should calulate surcharge for distance more than than 1000 meter', () => {
        expect(
            deliveryChargeCalculator(11, 1549, 1, new Date(2022, 1, 2, 14, 30))
        ).toBe(400);
    });

    it('should calulate surcharge if amount of items are more than 4', () => {
        expect(
            deliveryChargeCalculator(11, 1000, 7, new Date(2022, 1, 2, 14, 30))
        ).toBe(350);
    });

    it('should calulate surcharge if amount of items are more than 12', () => {
        expect(
            deliveryChargeCalculator(11, 1000, 14, new Date(2022, 1, 2, 14, 30))
        ).toBe(820);
    });

    it('should calulate surcharge during peak hour', () => {
        expect(
            deliveryChargeCalculator(
                11,
                1000,
                14,
                new Date(2023, 0, 27, 15, 30)
            )
        ).toBe(984);
    });

    it('should calculate delivery fee to be 0 for cart value more than or equal to 100 euro', () => {
        expect(
            deliveryChargeCalculator(
                100,
                1000,
                14,
                new Date(2023, 0, 27, 15, 30)
            )
        ).toBe(0);
    });

    it('should reutrn NaN for delivery fee more than 15 euro', () => {
        expect(
            deliveryChargeCalculator(
                90,
                5000,
                14,
                new Date(2023, 0, 27, 15, 30)
            )
        ).toBe(NaN);
    });
});
