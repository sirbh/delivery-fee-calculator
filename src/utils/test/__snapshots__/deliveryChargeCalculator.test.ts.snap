// Vitest Snapshot v1

exports[`delivery charge calculator > snapshot 1`] = `
{
  "foo": "bar",
}
`;

exports[`delivery charge calculator test > snapshot 1`] = `
{
  "foo": "bar",
}
`;

exports[`suite name > snapshot 1`] = `
{
  "foo": "bar",
}
`;
